File: quote/Projects/PluginModel/template/Detail.vb
Imports System.ComponentModel
Imports System.Reflection

Namespace Template

    Public Class Detail
        Inherits Common.Detail

        Private _WireProperties As Common.WireProperties
        Private _ComponentProperties As Common.ComponentProperties

        Public Sub New(ByVal header As Header, ByVal product As Product)
            MyBase.New(product, product.UnitOfMeasure)
    
            Me.Header = header
            Me.IsWire = product.IsWire
            Me._ComponentProperties = PropertyFactory.CreateComponentProperties(Me)
            Me._WireProperties = PropertyFactory.CreateWireProperties(Me)
            Me.SequenceNumber = Me.Header.NextSequenceNumber
        End Sub

        <BrowsableAttribute(False)>
        Property Header As Header

        Public Sub UpdateComponentProperties(ByVal product As Product)
    
            MyBase.SetProduct(product)
            MyBase.SetPrivateUnitOfMeasure(product.UnitOfMeasure)
            
            Me._ComponentProperties = PropertyFactory.CreateComponentProperties(Me)
            
            If TypeOf Me._WireProperties Is DefaultWireProperties Then
                Dim w As DefaultWireProperties = Me._WireProperties
                w.PoundsPer1000Feet = product.CopperWeightPer1000Feet
            End If
            
        End Sub

        <BrowsableAttribute(False)>
        Public Overrides ReadOnly Property QuoteDetailProperties As Object
            Get
                If Me.IsWire Then
                    Return _WireProperties
                Else
                    Return _ComponentProperties
                End If
            End Get
        End Property

    End Class
End Namespace

File: quote/Projects/PluginModel/template/Detail.cs
namespace Model.Template
{
    using System;
    using System.ComponentModel;
    using System.Linq;
    
    using Model.Template.Ext;

    public class Detail : Common.Detail
    {
        private Common.WireProperties wireProperties;

        private Common.ComponentProperties componentProperties;
        
        public Detail(Header header, Product product) : base(product, product.UnitOfMeasure, 1)
        {
            this.Header = header;
            this.componentProperties = PropertyFactory.CreateComponentProperties(this);
            this.wireProperties = PropertyFactory.CreateWireProperties(this);
            this.SequenceNumber = this.Header.NextSequenceNumber;
            
            this.AddChildProperty(this.componentProperties);
            this.AddChildProperty(this.wireProperties);
        }

        [BrowsableAttribute(false)]
        public Header Header { get; set; }

        [BrowsableAttribute(false)]
        public override Common.ISavableProperties QuoteDetailProperties 
        {
            get 
            {
                if (this.IsWire) 
                {
                    return this.wireProperties;
                } 
                else 
                {
                    return this.componentProperties;
                }
            }
        }

        public void UpdateComponentProperties(Product product)
        {
            this.SetProduct(product);
            this.SetPrivateUnitOfMeasure(product.UnitOfMeasure);

            this.componentProperties = PropertyFactory.CreateComponentProperties(this);

            if (this.wireProperties is DefaultWireProperties) 
            {
                DefaultWireProperties w = this.wireProperties as DefaultWireProperties;
                w.PoundsPer1000Feet = product.CopperWeightPer1000Feet;
            }
        }
    }
}
